### PR TITLE
Tarred dataset for text normalization model

### PR DESCRIPTION
**Pull request by ekmb**

### CODE REVIEW

File: nemo/collections/nlp/data/text_normalization/decoder_dataset.py [hunk 1]
```diff
@@ -307,3 +427,127 @@ def __init__(
...
+                available in the tarred dataset, which are permanently pre-allocated and never changed at runtime.
+                The benefit of replication is that it allows each node to sample data points from the entire
+                dataset independently of other nodes, and reduces dependence on value of `shuffle_n`.
+                Note: Replicated strategy allows every node to sample the entire set of available tarfiles,
+                and therefore more than one node may sample the same tarfile, and even sample the same
+                data points! As such, there is no assured guarantee that all samples in the dataset will be
+                sampled at least once during 1 epoch.
+        global_rank (int): Worker rank, used for partitioning shards. Defaults to 0.
+        world_size (int): Total number of processes, used for partitioning shards. Defaults to 1.
+        reverse_lang_direction (bool): When True, swaps the source and target directions when returning minibatches.
```
Comment (yzhang123) [inline nemo/collections/nlp/data/text_normalization/decoder_dataset.py:472]: missing arg

File: nemo/collections/nlp/data/text_normalization/decoder_dataset.py [hunk 2]
```diff
@@ -307,3 +427,127 @@ def __init__(
...
+        prepend_id (int): Prepends the specificed token id to the start of every source sentence. Defaults to None.
```
Comment (yzhang123) [inline nemo/collections/nlp/data/text_normalization/decoder_dataset.py:473]: missing arg

File: nemo/collections/nlp/data/text_normalization/decoder_dataset.py [hunk 3]
```diff
@@ -307,3 +427,127 @@ def __init__(
...
+    See the WebDataset documentation for more information about accepted data and input formats.
+    If using multiple processes the number of shards should be divisible by the number of workers to ensure an
+    even split among workers. If it is not divisible, logging will give a warning but training will proceed.
+    Additionally, please note that the len() of this DataLayer is assumed to be the number of tokens
+    of the text data. An incorrect manifest length may lead to some DataLoader issues down the line.
+    Args:
+        text_tar_filepaths: Either a list of tokenized text tarball filepaths, or a
+            string (can be brace-expandable).
+        metadata_path (str): Path to the metadata manifest.
+        encoder_tokenizer: Autokenizer wrapped BPE tokenizer model, such as YTTM
```
Comment (yzhang123) [inline nemo/collections/nlp/data/text_normalization/decoder_dataset.py:454]: missing arg

File: nemo/collections/nlp/data/text_normalization/decoder_dataset.py [hunk 4]
```diff
@@ -307,3 +427,127 @@ def __init__(
...
+        decoder_tokenizer: Autokenizer wrapped BPE tokenizer model, such as YTTM
```
Comment (yzhang123) [inline nemo/collections/nlp/data/text_normalization/decoder_dataset.py:455]: missing arg

**Review by yzhang123 — CHANGES_REQUESTED** [review]

File: nemo/collections/nlp/models/duplex_text_normalization/duplex_decoder.py [hunk 5]
```diff
@@ -453,50 +469,80 @@ def setup_test_data(self, test_data_config: Optional[DictConfig]):
...
+                input_file=input_file,
+                tokenizer=self._tokenizer,
+                tokenizer_name=self.transformer_name,
+                mode=self.mode,
+                max_len=cfg.get('max_decoder_len', self._tokenizer.model_max_length),
+                decoder_data_augmentation=cfg.get('decoder_data_augmentation', False)
+                if data_split == "train"
+                else False,
+                lang=self.lang,
+                do_basic_tokenize=cfg.do_basic_tokenize,
```
Comment (yzhang123) [inline nemo/collections/nlp/models/duplex_text_normalization/duplex_decoder.py:520]: please set explicitly do_tokenize=True

**Review by yzhang123 — CHANGES_REQUESTED** [review]

File: nemo/collections/nlp/data/text_normalization/decoder_dataset.py [hunk 6]
```diff
@@ -220,6 +282,64 @@ def __getitem__(self, idx):
...
+                batch['semiotic_class_id'] = classes
+
+                labels = self.tokenizer(targets, padding=True)['input_ids']
+                batch['decoder_input_ids'] = np.insert(
+                    [x[:-1] for x in labels], 0, self.tokenizer.pad_token_id, axis=-1
+                )
+
+                # use LABEL_PAD_TOKEN_ID to disregard padded values for the loss calculations
+                batch['labels'] = [[x if x != 0 else constants.LABEL_PAD_TOKEN_ID for x in l] for l in labels]
+                batches.append(batch)
```
Comment (yzhang123) [inline nemo/collections/nlp/data/text_normalization/decoder_dataset.py:337]: batch['attention_mask'] is missing?

File: nemo/collections/nlp/data/text_normalization/decoder_dataset.py [hunk 7]
```diff
@@ -220,6 +282,64 @@ def __getitem__(self, idx):
     def __len__(self):
         return len(self.examples)
 
+    def batchify(self, batch_size: int):
+        """
+        Creates a batch
```
Comment (yzhang123) [inline nemo/collections/nlp/data/text_normalization/decoder_dataset.py:287]: can you use DataCollatorForSeq2Seq here too, and pass model as argument?
And by creating Dataloader to help you generate the batch keys needed? Since these could change with different model?

**Review by yzhang123 — CHANGES_REQUESTED** [review]

**Review by yzhang123 — APPROVED** [review]

**Review by yzhang123 — COMMENTED** [review]
could you please add mini example to ci? @ekmb 
rest looks good, thanks!

**Review by yzhang123 — APPROVED** [review]

**Review by yzhang123 — APPROVED** [review]

**Comment by ekmb** [comment]
closing in favor of https://github.com/NVIDIA/NeMo/pull/2826
